Add parser tests part 1

mmulholla opened the pull request:
### What does this PR do?
Adds tests which use the parser api to verify, create. and edit valid devfile.yaml files.

Current support : 
Commands: Exec and Composite
Components: Container and Volume

### What issues does this PR fix or reference?
https://github.com/devfile/api/issues/215


### Is your PR tested? Consider putting some instruction how to test your changes
Yes. Readme includes instruction for how to modify tests

maysunfaisal commented:
@mmulholla make a temp commit and print `git status -s` in the CI workflow to see what files are in the git status thats causing it to fail

mmulholla commented:
I added a change to the build order to enable the build to pass. go fmt was failing because the pull request was running go tests and then checking go fmt. The go test creates a tmp directory which it leaves populated, then go fmt throws an error because of the tmp directory existing. I changed the build order so it checks format and then runs the tests and the build then passes.

I also tweeked build output after running go fmt to output the error message from go fmt - without thus i would never have found this issue!!

yangcao77 commented on tests/src/tests/README.md:
```diff
@@ -0,0 +1,231 @@
+    * Parser functions covered:
+        * Read an existing devfile.
+        * Write a new devfile.
+        * Modify Content of a devfile.
```
```suggestion
        * Modify content of a devfile.
```

yangcao77 commented on tests/src/tests/README.md:
```diff
@@ -0,0 +1,231 @@
+    * Composite
+* Components 
+    * Container
+    * Volume
```
since https://github.com/devfile/library/pull/48 has been merged, plugin component with uri defined  is also supported

yangcao77 commented on tests/src/tests/README.md:
````diff
@@ -0,0 +1,231 @@
+1. Go to directory tests/src/tests
+1. Run ```go test``` or ```go test -v```
+1. The test creates the following files: 
+    1. ```./tmp/test.log``` contains log output from the tests.
````
the `tmp` folder should be added in gitignore

yangcao77 commented on tests/src/tests/README.md:
```diff
@@ -0,0 +1,231 @@
+
+* Each devfile is created in a schema structure.
+* Which attributes are set and the values used are randomized.
+* Once the schema structure is complete it is written in one of two ways.
```
why? explain in which situation each is being picked to used?

yangcao77 commented on tests/src/tests/README.md:
````diff
@@ -0,0 +1,231 @@
+        * ```func create<command-type>Command() *schema.<command-type>```
+            * creates the command object and calls ```set<command-type>CommandValues``` to add attribute values
+            * for example see: ```func createExecCommand(execCommand *schema.ExecCommand)```
+        * ```func set<command-type>CommandValues(project-sourceProject *schema.<project-source>)```
````
suggest to switch order of `set<command-type>CommandValues` and `create<command-type>Command()`
should explain it before reference it

yangcao77 commented on tests/src/tests/README.md:
````diff
@@ -0,0 +1,231 @@
+            * main entry point for a test to add a command
+            * maintains the array of commands in the schema structure
+            * calls generateCommand() 
+        * ```func generateCommand(command *schema.Command, genericCommand *GenericCommand)```
````
same here, suggest to explain `generateCommand` before reference it under `addCommand`

yangcao77 commented on tests/src/tests/README.md:
````diff
@@ -0,0 +1,231 @@
+        * ```func (devfile *TestDevfile) UpdateCommand(command *schema.Command) error```
+            * includes logic to call set<commad-type>CommandValues for each commandType.
+        * ```func (devfile TestDevfile) VerifyCommands(parserCommands []schema.Command) error```
+            * Includes logic to compare the array of commands obtained from the parser with those created by the test. if the compare fails:
````
be consistent
```suggestion
            * includes logic to compare the array of commands obtained from the parser with those created by the test. If the compare fails:
```

yangcao77 commented on tests/src/tests/README.md:
```diff
@@ -0,0 +1,231 @@
+        * Includes code to get object from the paser and verify their content.
+        * For commands code is required to: 
+            1. Retrieve each command from the parser
+            1. Use command id to obtain the GenericCommand object which matches
```
since defined as `Id` in schema
```suggestion
            1. Use command Id to obtain the GenericCommand object which matches
```

yangcao77 commented on tests/src/tests/README.md:
```diff
@@ -0,0 +1,231 @@
+        * For commands code is required to: 
+            1. Retrieve each command from the parser
+            1. Use command id to obtain the GenericCommand object which matches
+            1. compare the command structure returned by the parser with the command structure saved in the GenericCommand object.
```
To be consistent
```suggestion
            1. Compare the command structure returned by the parser with the command structure saved in the GenericCommand object.
```

yangcao77 commented on tests/src/tests/README.md:
````diff
@@ -0,0 +1,231 @@
+    * ```func (devfile TestDevfile) EditCommands() error```
+        * Specific to command objects.
+            1. Ensure devfile is written to disk
+            1. use parser to read devfile and get all command object
````
```suggestion
            1. Use parser to read devfile and get all command object
```

yangcao77 commented on tests/src/tests/README.md:
````diff
@@ -0,0 +1,231 @@
+            1. use parser to read devfile and get all command object
+            1.  for each command call:
+                *  ```func (devfile *TestDevfile) UpdateCommand(command *schema.Command) error``` 
+            1. when all commands have been updated, use parser to write the updated devfile to disk
````
```suggestion
            1. When all commands have been updated, use parser to write the updated devfile to disk
```

yangcao77 commented on tests/src/tests/README.md:
```diff
@@ -0,0 +1,231 @@
+        * Specific to command objects.
+            1. Ensure devfile is written to disk
+            1. use parser to read devfile and get all command object
+            1.  for each command call:
```
```suggestion
            1.  For each command call:
```

yangcao77 commented on tests/src/tests/README.md:
````diff
@@ -0,0 +1,231 @@
+    * See also
+        *```func Test_<string>ExecCommand(t *testing.T)``` 
+        *```func Test_MultiCommand(t *testing.T)```
+        *```func Test_Everything(t *testing.T)```
````
Need space after `*` 
```suggestion
        * ```func Test_<string>ExecCommand(t *testing.T)``` 
        * ```func Test_MultiCommand(t *testing.T)```
        * ```func Test_Everything(t *testing.T)```
```

yangcao77 commented on tests/src/tests/README.md:
````diff
@@ -0,0 +1,231 @@
+        *```func Test_<string>ExecCommand(t *testing.T)``` 
+        *```func Test_MultiCommand(t *testing.T)```
+        *```func Test_Everything(t *testing.T)```
+    * add logic to ```func runTest(testContent TestContent, t *testing.T)``` includes logic
````
```suggestion
    * Adds logic to ```func runTest(testContent TestContent, t *testing.T)``` which includes logic
```

yangcao77 commented on tests/src/tests/README.md:
````diff
@@ -0,0 +1,231 @@
+        *```func Test_MultiCommand(t *testing.T)```
+        *```func Test_Everything(t *testing.T)```
+    * add logic to ```func runTest(testContent TestContent, t *testing.T)``` includes logic
+        1. Add commands to the test
````
```suggestion
        1. Adds commands to the test
```

yangcao77 commented on tests/src/tests/README.md:
```diff
@@ -0,0 +1,231 @@
+        1. command-test-utils.AddCommand
+            1. command-test-utils.GenerateCommand
+                1.  command-test-utils.createExecCommand
+                    1. command-test-utils.setExecCommandValues
```
check the indentation, otherwise the format will be incorrect. 
same same below
```suggestion
        1. command-test-utils.AddCommand
        1. command-test-utils.GenerateCommand
        1.  command-test-utils.createExecCommand
        1. command-test-utils.setExecCommandValues
```

mmulholla commented:
@yangcao77 Thank you for your comments, I have addressed each of them except:

"since #48 has been merged, plugin component with uri defined is also supported"

the section is about what is supported in the test which does not include that function yet.

yangcao77 commented on tests/src/tests/test_utils.go:
```diff
@@ -0,0 +1,324 @@
+	return GetRandomDecision(1, 1)
+}
+
+func GetRandomDecision(success int, failure int) bool {
```
how does this function works? why expect an int value for success and failure?
Besides, I would like to suggest having a simple comment for every function you added

mmulholla replied:
There are a lots of functions without comments. Is it OK if I add these in a subsequent pull request? The tests are working and would like to get them merged - this is only part 1.

mmulholla replied:
Have now added simple comments to all functions

yangcao77 commented on tests/src/tests/command_test_utils.go:
```diff
@@ -0,0 +1,256 @@
+
+	commandGroup.Kind = GetRandomGroupKind()
+	LogMessage(fmt.Sprintf("   ....... group Kind: %s", commandGroup.Kind))
+	commandGroup.IsDefault = GetBinaryDecision()
```
this might generate an invalid devfile, since each command group can only have one default command. 
check the point 5 in commands validation: https://github.com/devfile/api/blob/master/pkg/validation/validation-rule.md

```
{build, run, test, debug}, each kind of group can only have one default command associated with it. If there are multiple commands of the same kind without a default, a warning will be displayed.
```

mmulholla replied:
At the moment the code only adds one group to any single command, so this is not currently a problem. I may update the code to add multiple in which case I can handle the requirement then.

yangcao77 replied:
Sounds good. might want to open an issue to track for multiple command tests

yangcao77 commented on tests/src/tests/command_test_utils.go:
```diff
@@ -0,0 +1,256 @@
+					if err == nil {
+						err = ioutil.WriteFile(parserFilename, c, 0644)
+					}
+					LogMessage(fmt.Sprintf("   .......marshall and write devfile %s", devfile.FileName))
```
for the log messages format, e.g. `   .......` and ` --> ERROR:` have to be manually passes in. suggest separate functions:
```
LogErrorMessage(){
testLogger.Println("--> ERROR: " + message)
}
```
and
```
LogInfoMessage(){
testLogger.Println("   ....... " + message)
}
```

mmulholla replied:
Sure, if that is a format you prefer, also its a minor improvement so I prefer to add it in part 2.

yangcao77 replied:
Can you also create an issue to track this as well? I just do not think it's a good idea to manually input the log format for every log message.

yangcao77 commented on tests/src/tests/component_test_utils.go:
```diff
@@ -0,0 +1,228 @@
+	}
+
+	generateComponent(&components[index], componentType)
+	devfile.SchemaDevFile.Components = components
```
I am confused with this function.
any reason `generateComponent` requires `&components` instead of just returning a component?
any reason need another copy of `devfile.SchemaDevFile.Components`?

can we just do 
```
devfile.SchemaDevFile.Components = append(devfile.SchemaDevFile.Components, generateComponent( componentType))
```

mmulholla replied:
Does append increment the size of a slice?

yangcao77 replied:
different from java array, you do not need to defined size for a slice in golang
you do not need to do `components = make([]schema.Component, currentSize+1)`, as long as you defined the slice as `var components []schema.Component`, you are free to use it.

mmulholla replied:
Done

yangcao77 commented on tests/src/tests/component_test_utils.go:
```diff
@@ -0,0 +1,228 @@
+	if GetBinaryDecision() {
+		containerComponent.VolumeMounts = AddVolume(GetRandomNumber(4))
+	} else {
+		containerComponent.Env = nil
```
should be `VolumeMounts`?

yangcao77 replied:
if it's `nil`, you can just leave it empty instead of setting it as nil. same as above

mmulholla replied:
yes, oops

yangcao77 commented on tests/src/tests/component_test_utils.go:
```diff
@@ -0,0 +1,228 @@
+	LogMessage("Enter VerifyComponents")
+	errorString := ""
+
+	// Compare entire array of commands
```
components*

mmulholla replied:
oops

yangcao77 commented on tests/src/tests/component_test_utils.go:
```diff
@@ -0,0 +1,228 @@
+			}
+		}
+		for _, component := range devfile.SchemaDevFile.Components {
+			if _, found := getSchemaComponent(parserComponents, component.Name); !found {
```
since `parserComponents` and `devfile.SchemaDevFile.Components` are already compared using `cmp.Equal`, this for loop is redundant

yangcao77 replied:
rechecked,  the loop is executed if the two slices do not match. my bad

mmulholla replied:
The cmp.Equal doesn't give an error if it fails, so you don't which component(s) was in error. The second loop only runs if the cmp fails and it does a component by component compare so it can narrow down the error to a specific component. Just helps with diagnosis if an error should happen.

yangcao77 commented on tests/src/tests/endpoint-test-utils.go:
```diff
@@ -0,0 +1,59 @@
+	return Exposures[GetRandomNumber(len(Exposures))-1]
+}
+
+var Protocols = [...]schema.EndpointProtocol{schema.HTTPEndpointProtocol, schema.WSEndpointProtocol, schema.TCPEndpointProtocol, schema.UDPEndpointProtocol}
```
should `wss` and `https` be included as well?

mmulholla replied:
At the time of writing the tests the parser was not using the schema with this update. If it is now I can extend the test.

yangcao77 replied:
yes, it's been merged in api repo as a part of https://github.com/devfile/api/pull/268

yangcao77 commented on tests/src/tests/command_test_utils.go:
```diff
@@ -0,0 +1,256 @@
+		}
+		devfile.replaceSchemaCommand(*parserCommand)
+	} else {
+		errorString += LogMessage(fmt.Sprintf(" ....... Command not found in test : %s", parserCommand.Id))
```
I would prefer to have `errorString` with type `[]string`
and do `errorString= append(errorString, errorMessage)` instead of using `+=`

mmulholla replied:
Sure append is new to me :-), but another I would like to keep for part 2.

mmulholla replied:
done

yangcao77 commented on tests/src/tests/test_utils.go:
```diff
@@ -0,0 +1,324 @@
+
+func GetGroupKinds() []schema.CommandGroupKind {
+	setRandSeed()
+	return []schema.CommandGroupKind{schema.BuildCommandGroupKind, schema.RunCommandGroupKind, schema.TestCommandGroupKind, schema.DebugCommandGroupKind}
```
this is a constant slice, no need to setRandSeed. and no need a separate function for the group kinds
suggest to have the slice defined at top of the file, as you've done for protocol and exposure in endpoints

mmulholla replied:
Yes, I created endpoint later than components - maybe my go skills were increasing :-)

mmulholla replied:
Done

yangcao77 commented on tests/src/tests/test_utils.go:
```diff
@@ -0,0 +1,324 @@
+	testDevfile := TestDevfile{}
+	testDevfile.SchemaDevFile = schema.Devfile{}
+	testDevfile.FileName = fileName
+	testDevfile.SchemaDevFile.SchemaVersion = "2.0.0"
```
Just want to point out that I have seen this schema version is hard coded to `2.0.0`, and the tests are relying on devfile structs defined in api repo(schema). However, the parser repo is going to support multiple schema versions, not only V2, but also future versions. 
I know this is the first PR (part1) for the test, we might also want to start considering how the test structure is going to be for future schema versions, i.e. which parts can be commonly used, and which parts are version specific.

mmulholla replied:
Yes, that is on my to do list, not sure yet how to best do it.

maysunfaisal replied:
I honestly think we should do that before merging this PR. Once this merged, we will keep adding tests and before we know it, it will be a lot of updates and  refactoring to accommodate future version schema changes.

yangcao77 commented on tests/src/tests/test_utils.go:
```diff
@@ -0,0 +1,324 @@
+				}
+				err = devObj.WriteYamlDevfile()
+				if err != nil {
+					LogMessage(fmt.Sprintf(" ..... ERROR: wriring devfile : %v", err))
```
again, I saw the formats for log messages are inconsistent. 
here is ` ..... ERROR: ` for error messages, and above are using ` --> ERROR`
and I saw some messages are with 5 dots, and some are with 6 dots, and with/without spaces in between. 

this is due to the log format is manually input for every message, and which is something we want to avoid

mmulholla replied:
Yes, that will help with consistency, part 2? :-)

mmulholla replied:
Done

yangcao77 commented on tests/src/tests/test_utils.go:
```diff
@@ -0,0 +1,324 @@
+		if commands != nil && len(commands) > 0 {
+			err = devfile.VerifyCommands(commands)
+		} else {
+			LogMessage(fmt.Sprintf("  No command found in %s : ", devfile.FileName))
```
```suggestion
			LogMessage(fmt.Sprintf("  No commands found in %s : ", devfile.FileName))
```

mmulholla replied:
lol, the patent attorney in me says that "command" is more accurate.

yangcao77 commented on tests/src/tests/test_utils.go:
```diff
@@ -0,0 +1,324 @@
+
+			err = ctx.SetAbsPath()
+			if err != nil {
+				LogMessage(fmt.Sprintf(" ..... ERROR: setting devfile path : %v", err))
```
should return `err` if it's not nil. 
same for other error checks

mmulholla replied:
The code in this case does return err.

yangcao77 commented on tests/src/tests/test_utils.go:
```diff
@@ -0,0 +1,324 @@
+		LogMessage(fmt.Sprintf("   .......marshall and write devfile %s", devfile.FileName))
+		c, err := yaml.Marshal(&(devfile.SchemaDevFile))
+
+		if err == nil {
```
IMO, we should check `err!= nil`, and return err. 
same for other error checks

mmulholla replied:
Same here, it does get returned but only at the end. - single entry single exit

mmulholla replied:
Changed all err ==nil to err != nil to hopefully look less weird, but I keep a single exit point.

mmulholla commented:
@yangcao77 I have hopefully addressed all of your comments to your satisfaction. Also thanks again for the thorough review and teaching me about append.

maysunfaisal reviewed (changes requested):
Some review comments

Can we also update the path `tests/src/tests`?

- Dockerfiles can be in `tests/examples/dockerfiles`
- tests code themselves can in `tests/<test type like integration/e2e>` & `tests/utils`
- Readme can in `tests/`

maysunfaisal commented on tests/src/tests/command_test_utils.go:
```diff
@@ -0,0 +1,256 @@
+		for _, command := range devfile.SchemaDevFile.Commands {
+			commands[index] = command
+			index++
+		}
```
lol I see you already addressed the append in the latest commit and i missed it by 15 mins 😆

maysunfaisal commented on tests/src/tests/test_utils.go:
```diff
@@ -0,0 +1,394 @@
+		c, err := yaml.Marshal(&(devfile.SchemaDevFile))
+
+		if err != nil {
+			LogErrorMessage(fmt.Sprintf("Marshall devfile %s : %v", devfile.FileName, err))
```
need to return here because this `err` variable is different from the `var err error` in the first line of the func. So in the end where you `return err` it will be empty when this if block is executed. Same for the if block below this

mmulholla replied:
I fixed this to make sure the error is included in return err.

maysunfaisal commented on tests/src/tests/test_utils.go:
```diff
@@ -0,0 +1,394 @@
+		LogInfoMessage(fmt.Sprintf("Use Parser to write devfile %s", fileName))
+		newDevfile, err := devfileData.NewDevfileData(devfile.SchemaDevFile.SchemaVersion)
+		if err != nil {
+			LogErrorMessage(fmt.Sprintf("Creating new devfile : %v", err))
```
need to return here because this `err` variable is different from the `var err error` in the first line of the func. So in the end where you `return err` it will be empty when this if block is executed. Same for the if block below this

maysunfaisal commented on tests/src/tests/test_utils.go:
```diff
@@ -0,0 +1,394 @@
+}
+
+type TestDevfile struct {
+	SchemaDevFile   schema.Devfile
```
can we rename this to something else? Its very confusing to read `TestDevfile.SchemaDevFile`. Maybe Data or Contents sounds ok?

mmulholla replied:
I honestly cannot think of a more descriptive name - going to punt in this one.

maysunfaisal commented on tests/src/tests/test_utils.go:
```diff
@@ -0,0 +1,394 @@
+		if err != nil {
+			LogErrorMessage(fmt.Sprintf("Creating new devfile : %v", err))
+		} else {
+			newDevfile.SetSchemaVersion(devfile.SchemaDevFile.SchemaVersion)
+
+			// add the commands to new devfile
+			for _, command := range devfile.SchemaDevFile.Commands {
+				newDevfile.AddCommands(command)
+			}
+			// add components to the new devfile
+			newDevfile.AddComponents(devfile.SchemaDevFile.Components)
+
+			ctx := devfileCtx.NewDevfileCtx(fileName)
+
+			err = ctx.SetAbsPath()
+			if err != nil {
+				LogErrorMessage(fmt.Sprintf("Setting devfile path : %v", err))
+			} else {
+				devObj := parser.DevfileObj{
+					Ctx:  ctx,
+					Data: newDevfile,
+				}
+				err = devObj.WriteYamlDevfile()
+				if err != nil {
+					LogErrorMessage(fmt.Sprintf("Writing devfile : %v", err))
+				} else {
+					devfile.SchemaParsed = false
+				}
+			}
+
+		}
```
I see a lot of redundancy here. We are creating a test devfile with `devfile *TestDevfile` which ultimately is the same as `newDevfile, err := devfileData.NewDevfileData(devfile.SchemaDevFile.SchemaVersion)` 

Why not create your test devfile with `newDevfile, err := devfileData.NewDevfileData(devfile.SchemaDevFile.SchemaVersion)` in the first place?

You are storing the test data in two different structs here at this moment

mmulholla replied:
The test is meant to test the library api and as a result it should be using every api available, the code my look cumbersome but that is just how tests are. You raise an interesting point though, there may be parts of the api I am not testing fully and I need to do a review of that - make sure I fully exercise all apis.

maysunfaisal commented on tests/src/tests/test_utils.go:
```diff
@@ -0,0 +1,394 @@
+func (devfile *TestDevfile) ParseSchema() error {
+
+	var err error
+	if !devfile.SchemaParsed {
+		LogInfoMessage(fmt.Sprintf("Parse and Validate %s : ", devfile.FileName))
+		devfile.ParsedSchemaObj, err = devfilepkg.ParseAndValidate(devfile.FileName)
+		if err != nil {
+			LogErrorMessage(fmt.Sprintf("From ParseAndValidate %v : ", err))
+		}
+		devfile.SchemaParsed = true
+	}
```
Not sure what the point of parsing is, when you're curating the devfile struct with fake/dummy data in `CreateDevfile`. So the flow basically is populate the devfile struct, write to file and unmarshall it back to structs in a diff variable.. I think this all can be avoided!

maysunfaisal replied:
Is this because you want to verify across two structs in the end?

mmulholla replied:
It can look like madness sometimes but when it comes to tests "that is the way".  
The concept is i create a schema structure with attributes which is now my reference.
I write that as a yaml two ways: using the yaml tool and the parser.
I then use the parser to read the the file, and then check the content the parser returns to me matches my reference content.  Under the covers the parse may do "stuff" which sometimes makes this redundant but  "that is the way" and at least the "stuff" is tested. 
Hope you like the Mandalorian humour :-).

maysunfaisal commented on tests/src/tests/command_test_utils.go:
```diff
@@ -0,0 +1,258 @@
+func (devfile *TestDevfile) UpdateCommand(parserCommand *schema.Command) error {
+
+	var errorString []string
+	testCommand, found := getSchemaCommand(devfile.SchemaDevFile.Commands, parserCommand.Id)
```
why are we doing this? You already have the command that you passed in to this function - `parserCommand`. The `getSchemaCommand()` basically loops thru all the commands and matches the id and returns the same command back..

mmulholla replied:
So here I am testing the paser api used to modify existing devfile conent. So first I create the schema structure as reference devfile content. Write that to disk, have the parser read it and then start updating. For updating I update the reference structure, use parser api to make the same update, get the full devfile back and make sure the updates are reflected.  Probably need to get some invalid updates going here to that is for the future.

maysunfaisal replied:
I get what you mean, but you are not doing anything with the `testCommand` in this func:

```
var err error
	testCommand, found := getSchemaCommand(devfile.SchemaDevFile.Commands, parserCommand.Id)
	if found {
		LogInfoMessage(fmt.Sprintf("Updating command id: %s", parserCommand.Id))
		if testCommand.Exec != nil {
			setExecCommandValues(parserCommand.Exec)
		} else if testCommand.Composite != nil {
			setCompositeCommandValues(parserCommand.Composite)
		}
		devfile.replaceSchemaCommand(*parserCommand)
	} else {
		err = errors.New(LogErrorMessage(fmt.Sprintf("Command not found in test : %s", parserCommand.Id)))
	}
	return err
```

mmulholla replied:
testCommand is used in the if and else statements. The logic is basically

1. Use the id from the command returned by the parser to find the equivalent command in the test.
2: if it is found:
    a. check the type of the command recorded in the test and set appropriate update values in the command returned by the parser
   b. replace the command stored in the test with the new version to be given to the parser.
3. If it is not found we have a problem.

on return the parser api is use to write the updated parser command.

I could do this any number of ways but the outcome needs to be the same.

maysunfaisal commented on tests/src/tests/command_test_utils.go:
```diff
@@ -0,0 +1,258 @@
+		}
+		devfile.replaceSchemaCommand(*parserCommand)
+	} else {
+		errorString = append(errorString, LogErrorMessage(fmt.Sprintf("Command not found in test : %s", parserCommand.Id)))
```
Not sure what the point of having an `errorString []string` is here. We are not looping to store multiple errors.

mmulholla replied:
Fixed it.

maysunfaisal commented on tests/src/tests/command_test_utils.go:
```diff
@@ -0,0 +1,258 @@
+}
+
+// Use the specified command to replace the command in the schema structure with the same Id.
+func (devfile TestDevfile) replaceSchemaCommand(command schema.Command) {
+	for i := 0; i < len(devfile.SchemaDevFile.Commands); i++ {
+		if devfile.SchemaDevFile.Commands[i].Id == command.Id {
+			devfile.SchemaDevFile.Commands[i] = command
+			break
+		}
+	}
+}
```
I think if we clean up the test devfile structs like I mentioned earlier you can use the parser's updateCommand() func. This is basically duplicating effort and having two copies of the same thing - one in test and one in parser

mmulholla replied:
Yes, just the way the test work, maintain a reference of devfile content to compare with what the parser returns.

maysunfaisal reviewed (changes requested):
Think test devfile logic can be cleaned up more. There are a few redundant codes in tests.

mmulholla commented on tests/src/tests/README.md:
```diff
@@ -0,0 +1,235 @@
+
+The tests use the go language and are intended to test every apsect of the parser for every schema attribute. Some basic aspects of the tests:
+
+* A first test (parser_v200_schema_test.go) feeds pre-created devfiles to the parser to ensure the parser can parse all attribues and return an approproate error when the devfile contains an error.
```
Added a note to say the test was not available yet - needs updating as a result of schema changes

maysunfaisal replied:
If its not there, should we really merge this information? Dont want others to be confused whoever is reading it

mmulholla commented on tests/src/tests/test_utils.go:
```diff
@@ -0,0 +1,324 @@
+const tmpDir = "./tmp/"
+const logErrorOnly = false
+const logFileName = "test.log"
+const logToFileOnly = true // If set to false the log output will also be output to the console
```
Changed the constant format. 
Customizing the settings is using ENV or similar is for the future.

mmulholla commented:
Guys, thank you for the thorough reviews. I have hopefully responded to all of your comments and addressed most of them.
Please now approve. This is a first pass at adding some parser api tests, at this stage they don't need to be perfect (if tests ever need to be perfect) and they are a work in progress. I know there is a bunch of stuff that is not complete, there are better ways of doing some things, and they are still just touching a tiny percentage of the api, but it is a starting point and an important one. The tests also work.

mmulholla commented:
@maysunfaisal @yangcao77 Are you OK to approve now?

maysunfaisal commented on tests/utils/test_utils.go:
```diff
@@ -52,37 +55,23 @@ func init() {
-	testFileExtension := filepath.Ext(testFile)
-	subdir := testFile[0 : len(testFile)-len(testFileExtension)]
-	return CreateTempDir(subdir)
+	testError(true)
+	testError1(true)
```
Are these func and their calls supposed to be removed?

mmulholla replied:
oops.

maysunfaisal commented on tests/utils/project_test_utils.go:
```diff
@@ -0,0 +1,10 @@
+	schema "github.com/devfile/api/pkg/apis/workspaces/v1alpha2"
+)
+
+type GenericProject struct {
+	Name    string
+	Project *schema.Project
+}
```
pls delete this file and struct, its not being used anywhere

mmulholla replied:
Will remove the file. I have a branch waiting on this which implements this fully.

maysunfaisal commented on tests/utils/test_utils.go:
```diff
@@ -0,0 +1,433 @@
+}
+
+// EditCommands modifies random attributes for each of the commands in the devfile.
+func (devfile TestDevfile) EditCommands() error {
```
should this be in command utils file?

mmulholla replied:
I prefer where it is because the use the function makes of the parser, wanted to keep the command_utils pure command stuff.

maysunfaisal commented on tests/utils/test_utils.go:
```diff
@@ -0,0 +1,433 @@
+}
+
+// EditComponents modifies random attributes for each of the components in the devfile.
+func (devfile TestDevfile) EditComponents() error {
```
should this be in component utils file?

mmulholla replied:
I prefer where it is because the use the function makes of the parser, wanted to keep the component_utils pure component stuff.

maysunfaisal commented on tests/utils/endpoint-test-utils.go:
```diff
@@ -0,0 +1,63 @@
+	return Exposures[GetRandomNumber(len(Exposures))-1]
+}
+
+//var Protocols = [...]schema.EndpointProtocol{schema.HTTPEndpointProtocol, schema.HTTPSEndpointProtocol, schema.WSEndpointProtocol, schema.WSSEndpointProtocol, schema.TCPEndpointProtocol, schema.UDPEndpointProtocol}
```
remove comment

mmulholla replied:
It is a place holder for when I can run against the latest schema, at which point it will be uncomented.

maysunfaisal commented on tests/utils/endpoint-test-utils.go:
```diff
@@ -0,0 +1,63 @@
+var Exposures = [...]schema.EndpointExposure{schema.PublicEndpointExposure, schema.InternalEndpointExposure, schema.NoneEndpointExposure}
+
+// GetRandomExposure returns a random exposure value
+func GetRandomExposure() schema.EndpointExposure {
```
func can be unexported, used only within the pkg

maysunfaisal commented on tests/utils/endpoint-test-utils.go:
```diff
@@ -0,0 +1,63 @@
+var Protocols = [...]schema.EndpointProtocol{schema.HTTPEndpointProtocol, schema.WSEndpointProtocol, schema.TCPEndpointProtocol, schema.UDPEndpointProtocol}
+
+// GetRandomProtocol returns a random protocol value
+func GetRandomProtocol() schema.EndpointProtocol {
```
func can be unexported, used only within the pkg

maysunfaisal commented on tests/utils/endpoint-test-utils.go:
```diff
@@ -0,0 +1,63 @@
+}
+
+// CreateEndpoints createa and returns a randon numebr of endpoints in a schema structure
+func CreateEndpoints() []schema.Endpoint {
```
func can be unexported, used only within the pkg

maysunfaisal commented on tests/utils/command_test_utils.go:
```diff
@@ -0,0 +1,253 @@
+)
+
+// AddEnv creates and returns a specifed number of env attributes in a schema structure
+func AddEnv(numEnv int) []schema.EnvVar {
```
func can be unexported, used only within the pkg

pls check for all funcs in this file.

maysunfaisal commented on tests/utils/component_test_utils.go:
```diff
@@ -0,0 +1,233 @@
+)
+
+// AddVolume returns volumeMounts in a schema structure based on a specified number of volumes
+func AddVolume(numVols int) []schema.VolumeMount {
```
func can be unexported, used only within the pkg

pls check for all funcs in this file.

mmulholla replied:
Done. Envisioned a time i the future when tests may get this specific and need these apis but unexported for now.

maysunfaisal commented on tests/utils/test_utils.go:
```diff
@@ -0,0 +1,433 @@
+		} else {
+			LogInfoMessage(fmt.Sprintf("No command found in %s : ", devfile.FileName))
+		}
+	}
+
+	if err == nil {
```
no point is another if block? You can have the logic in the same block

mmulholla replied:
Moved.

maysunfaisal commented on tests/utils/command_test_utils.go:
```diff
@@ -0,0 +1,253 @@
+			}
+
+		}
+		for _, command := range devfile.SchemaDevFile.Commands {
+			if _, found := getSchemaCommand(parserCommands, command.Id); !found {
+				errorString = append(errorString, LogErrorMessage(fmt.Sprintf("Command from test not returned by parser : %s ", command.Id)))
+			}
+		}
```
Any reason we are looping again over here? We just looped thru test devfile and parsed commands in the above for loop.

mmulholla replied:
The previous loop goes through all the commands returned by the parser and makes sure they are the same as those expected by test. This loop does a quick check that the test does not have commands that the parser did not return.

maysunfaisal commented on tests/utils/component_test_utils.go:
```diff
@@ -0,0 +1,233 @@
+				errorString = append(errorString, LogErrorMessage(fmt.Sprintf("Component from parser not known to test - id : %s ", component.Name)))
+			}
+		}
+		for _, component := range devfile.SchemaDevFile.Components {
+			if _, found := getSchemaComponent(parserComponents, component.Name); !found {
+				errorString = append(errorString, LogErrorMessage(fmt.Sprintf("Component from test not returned by parser : %s ", component.Name)))
+			}
+		}
```
Same as the command verify comment.

mmulholla replied:
The previous loop goes through all the components returned by the parser and makes sure they are the same as those expected by test. This loop does a quick check that the test does not have components that the parser did not return.

maysunfaisal reviewed (changes requested):
Few other minor nits and comments

mmulholla commented:
@maysunfaisal - all done now?

maysunfaisal reviewed (approved):
/lgtm
